src/main/angular/app/services/real_services/data_service_impl.service.ts

```diff
@@ -7,27 +7,38 @@ import {DataService} from '../data.service';
 import {ErrorMessage} from '../../../model/error_message';
 import {ProjectMetaData} from '../../../model/project_metadata';
 
-/** Service which actually retrieves data from the server. */
+/** Service which actually retrieves data from the server. Will cache graph data. */
 @Injectable()
 export class DataServiceImpl implements DataService {
-  constructor(private http: HttpClient) {}
+  /** Used for caching graph data during a page's lifecycle. Used so removing a project
+   * from the graph and adding it back will not fire off an unnecessary GET request */
+  private cache: {[id: string]: ProjectGraphData};
```

Comment:
When caching data, it's important to consider how long before the cache becomes out of date. 

For example, let's suppose that a user doesn't refresh their browser for a week (this happens WAY TOO OFTEN). What would the user see? In this case, they'd see a week old data. 

Consider adding a cache invalidation mechanism. If there's not a handy typescript or angular cache class, feel free to implement a lightweight one.

Reply:
+1